pkg/kube/krt/README.md

```diff
@@ -0,0 +1,88 @@
+  * `krt` can accept any object type, from any source, and handle them the same.
+* Provide high level abstractions.
+  * Controller authors can write simple transformation functions from `Input` -> `Output` (with dependencies); the framework handles all the state automatically.
+
+## Key Primitives
+
+The most important primitive provided is the `Collection` interface.
+This is basically an `Informer`, but not tied to Kubernetes.
+
+Currently, there are three ways to build a `Collection`:
+* Built from an `Informer` with `WrapClient` or `NewInformer`.
+* Statically configured with `NewStatic`.
+* Derived from other collections (more information on this below).
+
+Unlike `Informers`, these primitives work on arbitrary objects.
+However, these objects are expected to have some properties, depending on their usage.
+These are *not* expressed as generic constraints due to limitations in Go's type system.
+
+* Each object `T` must have a unique `Key[T]` (which is just a typed wrapper around `string`) that uniquely identifies the object.
+    Default implementations exist for Kubernetes objects, Istio `config.Config` objects, and `ResourceName() string` implementations.
+* `Equals(k K) bool` may be implemented to provide custom implementations to compare objects. Comparison is done to detect if changes were made.
+  Default implementations are available for Kubernetes and protobuf objects, and will fallback to `reflect.DeepEqual`.
+* A `Name`, `Namespace`, `Labels`, and `LabelSelector` may optionally be included, for use with filters (see below).
+
+## Derived Collections
+
+The core of the framework is in the ability to derive collections from others.
+
+In general, these are built by providing some `func(inputs...) outputs...`.
+While more could be expressed, there are currently three forms implemented.
+
+* `func() *O` via `NewSingleton`
+  * This generates a collection that has a single value. An example would be some global configuration.
+* `func(input I) *O` via `NewCollection`
+  * This generates a one-to-one mapping of input to output. An example would be a transformation from a `Pod` type to a generic `Workload` type.
+* `func(input I) []O` via `NewManyCollection`
+  * This generates a one-to-many mapping of input to output. An example would be a transformation from a `Service` to a _set_ of `Endpoint` types.
+
+The form used and input type only represent the _primary dependencies_, indicating the cardinality.
+Each transformation can additionally include an arbitrary number of dependencies, fetching data from other collections.
+
+For example, a simple `Singleton` example that keeps track of the number of `ConfigMap`s in the cluster:
+
+```go
+ConfigMapCount := krt.NewSingleton[int](func(ctx krt.HandlerContext) *int {
+    cms := krt.Fetch(ctx, ConfigMaps)
+    return ptr.Of(len(cms))
+})
+```
+
+The `Fetch` operation enables querying against other collections.
```

Comment:
So why isn't ConfigMaps a "primary dependency" of ConfigMapCount? Is it just because there's no many-to-one derivation form? Or is it that the caller shouldn't need to pass ConfigMaps in each time? 

IMO, this will be the first lesson to learn trying to use krt; figuring out when to create a primary dependency vs. an inner fetch

Reply:
If it was you would have  the function called once for an individual configmap. But we need all the configmaps, not just one.

I add a fairly long section with lots of examples going more in depth on this.

Reply:
Interesting; so if a "many-to-many" derivation form existed, would it make sense to use that instead of `Fetch`? Or is a "many-to-many" form explicitly avoided in favor of Fetching within a derived collection. Either is fine, just trying to understand what is intentional vs. just left out of the doc

Reply:
Well in this case we have only one output, a single integer. So its many-to-one relationship? might be misunderstanding

Reply:
Ah no I got my wires crossed. Your first response makes sense.